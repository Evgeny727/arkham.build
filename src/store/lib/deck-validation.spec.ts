import { beforeAll, describe, expect, it } from "vitest";
import type { StoreApi } from "zustand";
import type { Deck } from "@/store/schemas/deck.schema";
import { getMockStore } from "@/test/get-mock-store";
import {
  selectLocaleSortingCollator,
  selectLookupTables,
  selectMetadata,
} from "../selectors/shared";
import type { StoreState } from "../slices";
import { validateDeck } from "./deck-validation";
import { resolveDeck } from "./resolve-deck";

const tests = [
  // Basic, valid decks
  ["valid: no errors", "base_case"],
  ["valid: forbidden card of quantity 0", "forbidden_quantity_0"],
  ["valid: max XP Lola", "max_xp_lola"],
  // Deck size
  ["invalid: too few cards", "too_few_cards"],
  ["invalid: too many cards", "too_many_cards"],
  // Deck limit
  ["invalid: too many copies of a card", "deck_limit_exceeded"],
  [
    "invalid: too many copies of a card by subname",
    "deck_limit_exceeded_subname",
  ],
  ["invalid: too many copies of a card by name", "deck_limit_exceeded_name"],
  // Signatures
  ["invalid: no signature", "required_missing_signature"],
  ["invalid: no signature weakness", "required_missing_weakness"],
  ["valid: replacement signatures only", "required_replacements"],
  [
    "valid: signatures and replacement signatures",
    "required_replacements_in_addition",
  ],
  ["valid: advanced signatures", "required_advanced_only"],
  [
    "valid: advanced signature and a signature weakness (after challenge scenario)",
    "required_advanced",
  ],
  ["valid: all possible signatures", "required_all"],
  ["valid: signatures with correct quantity", "required_quantity"],
  ["invalid: signatures with incorrect quantity", "required_quantity_invalid"],
  ["invalid: missing replacement signature", "required_replacements_invalid"],
  ["invalid: missing advanced signature", "required_advanced_invalid"],
  // Random basic weaknesses
  ["valid: random basic weakness present", "rbw"],
  [
    "valid: second stage of multi-stage random basic weakness present",
    "rbw_multistage",
  ],
  ["invalid: missing random basic weakness", "rbw_invalid_missing"],
  // Off-class access
  ["valid: dunwich off-class access", "dunwich"],
  ["invalid: too many dunwich off-class cards", "dunwich_invalid"],
  ["valid: off-class access with faction selection", "faction_select"],
  [
    "invalid: off-class access with faction selection",
    "faction_select_invalid",
  ],
  ["valid: tag-based off-class access", "tag_based_access"],
  [
    "valid: tag-based access and versatile overlap",
    "tag_based_access_versatile",
  ],
  [
    "invalid: tag-based access and versatile overlap",
    "tag_based_access_versatile_invalid",
  ],
  ["invalid: tag-based off-class access", "tag_based_access_invalid"],
  ["valid: trait-based off-class access", "trait_based_access"],
  ["invalid: trait-based off-class access", "trait_based_access_invalid"],
  // Customizable cards
  ["valid: customizable level in investigator range", "customizable_level"],
  ["invalid: customizable level exceeded", "customizable_level_exceeded"],
  [
    "invalid: customizable level below investigator range",
    "customizable_level_below",
  ],
  // At least constraints
  ["valid: 'at least x cards of y factions' requirement", "atleast_factions"],
  [
    "invalid: 'at least x cards of y factions' requirement",
    "atleast_factions_invalid",
  ],
  // Deck size adjustments
  ["valid: Forced Learning with correct deck size", "forced_learning"],
  ["valid: Underworld Market with correct deck size", "underworld_market"],
  // Ancestral Knowledge
  [
    "valid: Ancestral Knowledge with at least 10 skill cards",
    "ancestral_knowledge",
  ],
  [
    "invalid: Ancestral Knowledge with less than 10 skill cards",
    "ancestral_knowledge_invalid",
  ],
  // Covenants
  ["valid: one Covenant", "covenant"],
  ["invalid: more than one Covenant", "covenant_invalid"],
  // Mandy Thompson
  ["valid: Mandy Thompson with deck size selected", "mandy"],
  ["valid: Mandy Thompson taboo forces deck size of 50", "mandy_taboo"],
  ["invalid: Mandy Thompson with too few cards", "mandy_too_few_cards"],
  ["invalid: Mandy Thompson with too many cards", "mandy_too_many_cards"],
  [
    "invalid: Mandy Thompson with too few signatures for deck size",
    "mandy_signature_count_invalid",
  ],
  // Underworld Support
  ["valid: Underworld Support with only singletons", "underworld_support"],
  [
    "valid: Underworld Support with more than one copy of a a weakness",
    "underworld_support_weaknesses",
  ],
  [
    "invalid: Underworld Support invalid deck size",
    "underworld_support_invalid_size",
  ],
  [
    "invalid: Underworld Support with more than one copy of a card",
    "underworld_support_invalid_deck_limit",
  ],
  [
    "invalid: Underworld Support with more than one copy of a myriad card",
    "underworld_support_invalid_myriad",
  ],
  // Honed Instinct
  [
    "valid: three copies Honed Instinct with Impulse Control unlocked",
    "honed_instinct",
  ],
  [
    "invalid: three copies Honed Instinct with Impulse Control not unlocked",
    "honed_instinct_invalid",
  ],
  // Versatile
  ["valid: Versatile with added off-class card", "versatile"],
  ["invalid: Versatile with too many off-class cards", "versatile_invalid"],
  // || Jenny
  ["valid: || Jenny Barnes with 10 off-class Talents", "parallel_jenny"],
  [
    "invalid: || Jenny Barnes with more than 10 off-class Talents",
    "parallel_jenny_invalid_limit",
  ],
  [
    "invalid: || Jenny Barnes with permanent Talents",
    "parallel_jenny_invalid_forbidden",
  ],
  // || Roland (front)
  ["valid: || Roland (front) with three Directives", "parallel_roland"],
  [
    "invalid: || Roland (front) with too many Directives",
    "parallel_roland_invalid",
  ],
  // || Wendy (front)
  ["valid: || Wendy (front) with Tidal Memento", "parallel_wendy_front"],
  [
    "invalid: || Wendy (front) without Tidal Memento",
    "parallel_wendy_front_invalid",
  ],
  // || Wendy (back)
  ["valid: || Wendy (back) valid for selected option", "parallel_wendy"],
  [
    "invalid: || Wendy (back) invalid for selected option",
    "parallel_wendy_invalid",
  ],
  // Precious Memento
  ["valid: Precious Memento with correct limits", "precious_memento"],
  [
    "invalid: Too many copies of a Precious Memento",
    "precious_memento_invalid",
  ],
  // On Your Own
  ["valid: On Your Own without allies in deck", "on_your_own"],
  ["valid: On Your Own with story ally in deck", "on_your_own_story_ally"],
  [
    "valid: On Your Own with Summoned Servitor, upgraded to lose ally slot",
    "on_your_own_summoned_servitor",
  ],
  ["invalid: On Your Own with ally in deck", "on_your_own_invalid"],
  [
    "invalid: On Your Own with Summoned Servitor occupying an ally slot",
    "on_your_own_summoned_servitor_invalid",
  ],
  // Transformed investigator
  ["valid: transformed investigator (Y'thian)", "transformed_investigator"],
  // Marie (promo)
  ["valid: Marie (promo) with promo signatures", "promo_marie"],
  ["invalid: Marie (promo) with missing signature", "promo_marie_invalid"],
  // Eldritch Brand
  ["valid: Eldritch Brand with a singleton attachment", "eldritch_brand"],
  [
    "invalid: Eldritch Brand with too many copies of an attachment",
    "eldritch_brand_invalid",
  ],
  // Lily Chen
  [
    "valid: Lily Chen with correct pairs of signatures and weaknesses",
    "lily_chen",
  ],
  [
    "invalid: Lily Chen with too few weaknesses",
    "lily_chen_invalid_too_few_weaknesses",
  ],
  [
    "invalid: Lily Chen with too many weaknesses",
    "lily_chen_invalid_too_many_weaknesses",
  ],
  // || Jim / Extra deck
  ["valid: extra deck with no error", "extra_deck"],
  ["invalid: extra deck with forbidden cards", "extra_deck_invalid_forbidden"],
  [
    "invalid: extra deck with too many copies of a card",
    "extra_deck_invalid_deck_limit",
  ],
  [
    "invalid: extra deck with too many cards",
    "extra_deck_invalid_size_exceeded",
  ],
  ["invalid: extra deck with too few cards", "extra_deck_invalid_size_below"],
  // Silas Marsh
  ["valid: Silas Marsh with cycle signatures", "silas_marsh"],
  [
    "valid: Silas Marsh with replacement signatures",
    "silas_marsh_replacements",
  ],
  // || Agnes
  ["valid: || Agnes with ignored duplicates", "parallel_agnes"],
  ["valid: || Agnes with card over deck_limit", "parallel_agnes_deck_limit"],
  [
    "invalid: || Agnes with card over deck_limit",
    "parallel_agnes_invalid_deck_limit",
  ],
  // Myriad
  ["valid: myriad cards with different subnames", "myriad_subname"],
  ["invalid: myriad cards with different subnames", "myriad_subname_invalid"],
  // Suzi
  ["valid: Suzi with valid 'at_least' constraint", "suzi"],
  [
    "invalid: Suzi with invalid 'at_least' constraint and forbidden card",
    "suzi_invalid",
  ],
];

function validate(store: StoreApi<StoreState>, deck: Deck) {
  const state = store.getState();

  const metadata = selectMetadata(state);
  const lookupTables = selectLookupTables(state);

  return validateDeck(
    resolveDeck(
      {
        sharing: state.sharing,
        metadata,
        lookupTables,
      },
      selectLocaleSortingCollator(state),
      deck,
    ),
    metadata,
    lookupTables,
  );
}

describe("Deck Validation", () => {
  let store: StoreApi<StoreState>;

  beforeAll(async () => {
    store = await getMockStore();
  });

  it.for(tests)("%s", async ([name, fileName]) => {
    const expectation = (name as string).startsWith("valid:");
    const deck = await import(
      `@/test/fixtures/decks/validation/${fileName}.json`
    );
    const result = validate(store, deck);
    expect(result.valid).toEqual(expectation);
    if (!expectation) expect(result.errors).toMatchSnapshot();
  });
});
