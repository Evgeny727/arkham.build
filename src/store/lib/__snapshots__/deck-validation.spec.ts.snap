// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Deck Validation > invalid: 'at least x cards of y factions' requirement 1`] = `
[
  {
    "details": {
      "count": "(2 / 3)",
      "error": "You must have at least 7 cards from 3 different factions",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: || Agnes with card over deck_limit 1`] = `
[
  {
    "details": [
      {
        "code": "06201",
        "limit": 2,
        "quantity": 3,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: || Jenny Barnes with more than 10 off-class Talents 1`] = `
[
  {
    "details": {
      "count": "(11 / 10)",
      "error": "Too many off-class cards.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: || Jenny Barnes with permanent Talents 1`] = `
[
  {
    "details": [
      {
        "code": "60531",
        "real_name": "Déjà Vu",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: || Roland (front) with too many Directives 1`] = `
[
  {
    "details": [
      {
        "code": "90025",
        "quantity": 4,
        "required": 3,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: || Wendy (back) invalid for selected option 1`] = `
[
  {
    "details": {
      "count": 35,
      "countRequired": 30,
      "target": "slots",
    },
    "type": "TOO_MANY_CARDS",
  },
  {
    "details": [
      {
        "code": "07028",
        "real_name": "Faustian Bargain",
        "target": "slots",
      },
      {
        "code": "07032",
        "real_name": "Promise of Power",
        "target": "slots",
      },
      {
        "code": "07112",
        "real_name": "Stirring Up Trouble",
        "target": "slots",
      },
      {
        "code": "07113",
        "real_name": "Blasphemous Covenant",
        "target": "slots",
      },
      {
        "code": "07226",
        "real_name": "Armageddon",
        "target": "slots",
      },
      {
        "code": "07227",
        "real_name": "Eye of Chaos",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: || Wendy (front) without Tidal Memento 1`] = `
[
  {
    "details": [
      {
        "code": "90038",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: Ancestral Knowledge with less than 10 skill cards 1`] = `
[
  {
    "details": {
      "count": "(8 / 10)",
      "error": "Deck must have at least 10 skill cards.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: Eldritch Brand with too many copies of an attachment 1`] = `
[
  {
    "details": [
      {
        "code": "05322",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: Lily Chen with too few weaknesses 1`] = `
[
  {
    "details": [
      {
        "code": "08015",
        "quantity": 1,
        "required": 2,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: Lily Chen with too many weaknesses 1`] = `
[
  {
    "details": [
      {
        "code": "08015",
        "quantity": 2,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: Mandy Thompson with too few cards 1`] = `
[
  {
    "details": {
      "count": 39,
      "countRequired": 40,
      "target": "slots",
    },
    "type": "TOO_FEW_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: Mandy Thompson with too few signatures for deck size 1`] = `
[
  {
    "details": [
      {
        "code": "06008",
        "quantity": 3,
        "required": 2,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: Mandy Thompson with too many cards 1`] = `
[
  {
    "details": {
      "count": 41,
      "countRequired": 40,
      "target": "slots",
    },
    "type": "TOO_MANY_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: Marie (promo) with missing signature 1`] = `
[
  {
    "details": [
      {
        "code": "05018",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: On Your Own with Summoned Servitor occupying an ally slot 1`] = `
[
  {
    "details": [
      {
        "code": "09080",
        "real_name": "Summoned Servitor",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: On Your Own with ally in deck 1`] = `
[
  {
    "details": [
      {
        "code": "03198",
        "real_name": "Madame Labranche",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: Suzi with invalid 'at_least' constraint and forbidden card 1`] = `
[
  {
    "details": {
      "count": "(4 / 5)",
      "error": "You must have at least 7 cards from each class",
    },
    "type": "INVALID_DECK_OPTION",
  },
  {
    "details": [
      {
        "code": "02158",
        "real_name": "Charisma",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: Too many copies of a Precious Memento 1`] = `
[
  {
    "details": [
      {
        "code": "08114",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: Underworld Support invalid deck size 1`] = `
[
  {
    "details": {
      "count": 29,
      "countRequired": 28,
      "target": "slots",
    },
    "type": "TOO_MANY_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: Underworld Support with more than one copy of a card 1`] = `
[
  {
    "details": [
      {
        "code": "06117",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: Underworld Support with more than one copy of a myriad card 1`] = `
[
  {
    "details": [
      {
        "code": "06026",
        "limit": 1,
        "quantity": 3,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: Versatile with too many off-class cards 1`] = `
[
  {
    "details": {
      "count": "(2 / 1)",
      "error": "Too many off-class cards for Versatile.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: customizable level below investigator range 1`] = `
[
  {
    "details": {
      "count": "(7 / 5)",
      "error": "You cannot have more than 5 level 0 Guardian cards",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: customizable level exceeded 1`] = `
[
  {
    "details": [
      {
        "code": "09021",
        "real_name": "Hunter's Armor",
        "target": "slots",
      },
      {
        "code": "09022",
        "real_name": "Runic Axe",
        "target": "slots",
      },
      {
        "code": "09040",
        "real_name": "Alchemical Distillation",
        "target": "slots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: extra deck with forbidden cards 1`] = `
[
  {
    "details": [
      {
        "code": "06285",
        "real_name": "The Black Cat",
        "target": "extraSlots",
      },
    ],
    "type": "FORBIDDEN",
  },
]
`;

exports[`Deck Validation > invalid: extra deck with too few cards 1`] = `
[
  {
    "details": {
      "count": 9,
      "countRequired": 10,
      "target": "extraSlots",
    },
    "type": "TOO_FEW_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: extra deck with too many cards 1`] = `
[
  {
    "details": {
      "count": 11,
      "countRequired": 10,
      "target": "extraSlots",
    },
    "type": "TOO_MANY_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: extra deck with too many copies of a card 1`] = `
[
  {
    "details": [
      {
        "code": "01018",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: missing advanced signature 1`] = `
[
  {
    "details": [
      {
        "code": "01007",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: missing random basic weakness 1`] = `
[
  {
    "details": [
      {
        "code": "01000",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: missing replacement signature 1`] = `
[
  {
    "details": [
      {
        "code": "01006",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: more than one Covenant 1`] = `
[
  {
    "details": {
      "count": "(2 / 1)",
      "error": "You cannot have more than one Covenant in your deck.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: myriad cards with different subnames 1`] = `
[
  {
    "details": [
      {
        "code": "06243",
        "limit": 3,
        "quantity": 4,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: no signature 1`] = `
[
  {
    "details": [
      {
        "code": "01006",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: no signature weakness 1`] = `
[
  {
    "details": [
      {
        "code": "01007",
        "quantity": 0,
        "required": 1,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: off-class access with faction selection 1`] = `
[
  {
    "details": {
      "count": "(11 / 10)",
      "error": "Too many off-class cards.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: signatures with incorrect quantity 1`] = `
[
  {
    "details": [
      {
        "code": "89003",
        "quantity": 2,
        "required": 3,
      },
    ],
    "type": "DECK_REQUIREMENTS_NOT_MET",
  },
]
`;

exports[`Deck Validation > invalid: tag-based access and versatile overlap 1`] = `
[
  {
    "details": {
      "count": "(2 / 1)",
      "error": "Too many off-class cards for Versatile.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: tag-based off-class access 1`] = `
[
  {
    "details": {
      "count": "(16 / 15)",
      "error": "You cannot have more than 15 level 0-1 Seeker and/or Mystic cards",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: three copies Honed Instinct with Impulse Control not unlocked 1`] = `
[
  {
    "details": [
      {
        "code": "09061",
        "limit": 2,
        "quantity": 3,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: too few cards 1`] = `
[
  {
    "details": {
      "count": 41,
      "countRequired": 42,
      "target": "slots",
    },
    "type": "TOO_FEW_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: too many cards 1`] = `
[
  {
    "details": {
      "count": 43,
      "countRequired": 42,
      "target": "slots",
    },
    "type": "TOO_MANY_CARDS",
  },
]
`;

exports[`Deck Validation > invalid: too many copies of a card 1`] = `
[
  {
    "details": [
      {
        "code": "01047",
        "limit": 2,
        "quantity": 3,
      },
      {
        "code": "07224",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: too many copies of a card by name 1`] = `
[
  {
    "details": [
      {
        "code": "09095",
        "limit": 2,
        "quantity": 4,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: too many copies of a card by subname 1`] = `
[
  {
    "details": {
      "count": 8,
      "countRequired": 30,
      "target": "slots",
    },
    "type": "TOO_FEW_CARDS",
  },
  {
    "details": [
      {
        "code": "10059",
        "limit": 1,
        "quantity": 2,
      },
      {
        "code": "04022",
        "limit": 2,
        "quantity": 3,
      },
      {
        "code": "08041",
        "limit": 1,
        "quantity": 2,
      },
    ],
    "type": "INVALID_CARD_COUNT",
  },
]
`;

exports[`Deck Validation > invalid: too many dunwich off-class cards 1`] = `
[
  {
    "details": {
      "count": "(6 / 5)",
      "error": "You cannot have more than 5 cards that are not Guardian or Neutral",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;

exports[`Deck Validation > invalid: trait-based off-class access 1`] = `
[
  {
    "details": {
      "count": "(11 / 10)",
      "error": "Too many off-class cards.",
    },
    "type": "INVALID_DECK_OPTION",
  },
]
`;
